### ISSUE TITLE
The result of `car-atom` cannot be used in grounded operations.

### ISSUE THREAD

**Issue by Necr0x0Der** [trueagi-io/hyperon-experimental#224]
The code
```
(: if (-> Bool Atom Atom Atom))
(= (if True $then $else) $then)
(= (if False $then $else) $else)
(= (plus $atom)
  (if (== $atom ())
      0
      (+ (car-atom $atom) (plus (cdr-atom $atom)))))
!(plus (1 2 3 4))
```
doesn't work now. But it works if we change the type of `CarAtomOp` from `Atom::expr([ARROW_SYMBOL, ATOM_TYPE_EXPRESSION, ATOM_TYPE_ATOM])` to `Atom::expr([ARROW_SYMBOL, ATOM_TYPE_EXPRESSION, ATOM_TYPE_UNDEFINED])` (in stdlib.rs).
It is not a bug per se, but it is not convenient. It can also be worked around:
```
(: if (-> Bool Atom Atom Atom))
(= (if True $then $else) $then)
(= (if False $then $else) $else)
(: as-number (-> Atom Number))
(= (as-number $x) $x)
(= (plus $atom)
  (if (== $atom ())
      0
      (+ (as-number (car-atom $atom)) (plus (cdr-atom $atom)))))
!(plus (1 2 3 4))
```
So, it is really a minor issue. However, `as-number` looks hacky/unsafe, e.g. `!(+ (as-number "S") 2)` raises an exception. We may want to be able to specify the type of `car-atom` in such a way that the type of `(car (1 2 3))` would be `Number`. Also, we may want to disallow `(: as-number (-> Atom Number))` (or we may want the type-checker to be able to validate correctness of `(as-number 1)` or `(as-number (car-atom (1 2)))`, and to detect incorrectness of `(as-number "S")` at least at runtime).

**Comment by vsbogd** [comment]
> But it works if we change the type of CarAtomOp from Atom::expr([ARROW_SYMBOL, ATOM_TYPE_EXPRESSION, ATOM_TYPE_ATOM]) to Atom::expr([ARROW_SYMBOL, ATOM_TYPE_EXPRESSION, ATOM_TYPE_ATOM]) (in stdlib.rs).

@Necr0x0Der, two identical types here, could you please take a look how type should be changed?

**Comment by vsbogd** [comment]
> Also, we may want to disallow (: as-number (-> Atom Number)) (or we may want the type-checker to be able to validate correctness of (as-number 1) or (as-number (car-atom (1 2))), and to detect incorrectness of (as-number "S") at least at runtime).

Not sure how do you propose to do this? Generally the type signature `(: foo (-> Atom Number))` looks valid. Do you mean type-checker could analyze the implementation of `as-number` and find out that there are know grounded functions which can convert types properly?

**Comment by Necr0x0Der** [comment]
> > But it works if we change the type of CarAtomOp from Atom::expr([ARROW_SYMBOL, ATOM_TYPE_EXPRESSION, ATOM_TYPE_ATOM]) to Atom::expr([ARROW_SYMBOL, ATOM_TYPE_EXPRESSION, ATOM_TYPE_ATOM]) (in stdlib.rs).
> 
> @Necr0x0Der, two identical types here, could you please take a look how type should be changed? Do you mean `Atom::expr([ARROW_SYMBOL, ATOM_TYPE_EXPRESSION, ATOM_TYPE_NUMBER])` ?

Thanks, fixed. It was `ATOM_TYPE_UNDEFINED`, but it was lost for some reason.

**Comment by Necr0x0Der** [comment]
> > Also, we may want to disallow (: as-number (-> Atom Number)) (or we may want the type-checker to be able to validate correctness of (as-number 1) or (as-number (car-atom (1 2))), and to detect incorrectness of (as-number "S") at least at runtime).
> 
> Not sure how do you propose to do this? Generally the type signature `(: foo (-> Atom Number))` looks valid. Do you mean type-checker could analyze the implementation of `as-number` and find out that there are know grounded functions which can convert types properly?

I mean that
```
(: foo (-> Number))
(= (foo) "S")
```
is type-checked only if static type-checking is on (and `(: = (-> $t $t Type))` is defined), but the interpreter doesn't type-check equality queries composed by itself. That is, while evaluating `(foo)` it composes `(= (foo) $x)`, which is successfully pattern-matched against `(= (foo) "S")`, although the type constraint is not satisfied. We may agree that this is the problem of static type checking (being switched off), but we may also agree that with switched off static type checking, runtime type error should be thrown.

**Comment by Necr0x0Der** [comment]
PR #483 changed the return type of `let`. Should we finally change the return type of `car-atom` and `cdr-atom`? The last example is still an issue, but it is a different issue.

**Comment by Necr0x0Der** [comment]
There is still a general issue with a more precise type for `car-atom` and `cdr-atom`, but I consider this specific issue as fixed with #493
